tensorboard/data/server/main.rs

```diff
@@ -125,3 +147,12 @@ fn init_logging(default_log_level: LevelFilter) {
     use env_logger::{Builder, Env};
     Builder::from_env(Env::default().default_filter_or(default_log_level.to_string())).init();
 }
+
+/// Locks stdin and reads it to EOF, then exits the process.
+fn die_after_stdin() {
+    let stdin = std::io::stdin();
+    let stdin_lock = stdin.lock();
+    for _ in stdin_lock.bytes() {}
```

Comment:
My buffer of comment went away after some GitHub page navigation :( 

Are you intending to use `stdin: PIPE` on the parent process side? It feels a bit too key logger like for my liking. Just wondering, if we manually pass a stream to the stdin and rely on it closing upon crash?

Reply:
> My buffer of comment went away after some GitHub page navigation :(

That sucks; sorry to hear that. :-(

Yep, `stdin=PIPE`. I’ve just finished a patch for this; it works great.

Not sure what you mean by keylogger-like? With `stdin=PIPE`, a new pipe
is created, and it’s never connected to a tty or other input source. We
just use it to send a single bit of metadata—the “close” signal.

In other words, from the _subprocess’s_ perspective it’s stdin, but it’s
not the _user’s_ stdin.

Is this about privacy? If TensorBoard wanted to log your keystrokes, we
could just do that, with no need for cloak and dagger.

Opening a pipe with input stream other than 0 (stdin) is harder and less
portable, which is why I’ve avoided it. Reading from it on the Rust side
requires using `unsafe`, since you’re taking an arbitrary int and
asserting that it corresponds to an fd over which you have exclusive
ownership, which there is no way to verify. And on the Python side you
need, like, `close_fds=False`, and you probably need to think more about
what Windows means by “file handle”… it seems more complicated, and
I don’t see a real benefit from it.

I am certainly open to alternatives. My criteria are (a) portable
without too much special casing, (b) doesn’t require special testing on
Windows, and (c) avoids `unsafe` in Rust if possible. I think this
approach does well on (a) and (c) and does okay on (b)—ideally I would
like to test this, but it’s simple enough that I’m crossing my fingers
that it just works.

Reply:
> Not sure what you mean by keylogger-like? With stdin=PIPE, a new pipe
is created, and it’s never connected to a tty or other input source. We
just use it to send a single bit of metadata—the “close” signal.

> In other words, from the subprocess’s perspective it’s stdin, but it’s
not the user’s stdin.

Understand that but what does it get `PIPE`ed to? In Node-ism, isn't this akin to process.stdin.pipe(subprocess.stdin)?

> Is this about privacy? If TensorBoard wanted to log your keystrokes, we
could just do that, with no need for cloak and dagger.

Yes, I don't care if TensorBoard can do it today or whatever. More complex system that spans multiple process is harder to introspect to understand that and that is the problem at stake.

Others
No, I don't mean you should use file descriptor but merely create a stream object that you use to detect the closedness from the Rust process just like you have written so. Since I am not asking you to open up side stream other than the default `stdin` and `stdout`, there is no point in discussing the unsafe Rust-ism, right?

Reply:
> In Node-ism, isn't this akin to process.stdin.pipe(subprocess.stdin)?

It’s not `process.stdin.pipe` (i.e., piping the parent process’s stdin),
no. It’s analogous to:

```node
const child_process = require("child_process");
p = child_process.spawn("cat"); // stdin=PIPE is default in Node
p.stdout.on("data", (data) => void console.log("from cat: " + data));
p.stdin.write("hey\n");
// prints: "from cat: hey"
p.stdin.destroy();
```

I’ve searched for a bit to try to find a pure-Node “create a pipe”
equivalent but have come up empty. The hypothetical Node would be
roughly equivalent to:

```node
// Create a new anonymous pipe with a read end and a write end connected
// to each other but not to anything else.
const [read, write] = newPipe();
read.pipe(subprocess.stdin);
```

The C example of what’s happening (ignoring error handling) is:

```c
int pipefd[2];
pipe(pipefd); // new anonymous pipe from `pipefd[1]` to `pipefd[0]`
pid_t child = fork();
if (child < 0) abort(); // error in fork
if (child > 0) {
    // we're the parent
    close(pipefd[0]); // don't need to read from the pipe
    write(pipefd[1], "hey\n", 4);
    close(pipefd[1]); // send eof
    waitpid(child, NULL, 0);
    puts("parent done");
} else {
    // we're the child
    close(pipefd[1]); // don't need to write to the pipe
    dup2(pipefd[0], 0); // replace child stdin with the pipe's output
    execl("/bin/sed", "sed", "s/^/in child: /", NULL);
}
```

The key bits are the call to `pipe(pipefd)`, which creates a new
_anonymous_ pipe, and `dup2(pipefd[0], 0)`, which replaces the child’s
stdin with the other end of the pipe.

Does this help answer “what does it get `PIPE`ed to”? I think the
simplest answer really is just “nothing”.

> No, I don't mean you should use file descriptor but merely create a
> stream object that you use to detect the closedness from the Rust
> process just like you have written so. Since I am not asking you to
> open up side stream other than the default `stdin` and `stdout`, there
> is no point in discussing the unsafe Rust-ism, right?

I don’t think I understand this. What does it mean to “create a stream
object” without “opening up a side stream”? Or, perhaps, what you mean
is what’s already happening at that `pipe(pipefd)`.

Some more references, in case they help things click:

  - The [`Popen`] constructor says: “`PIPE` indicates that a *new pipe*
    to the child should be created […] With the default settings of
    `None`, no redirection will occur” (emphasis mine).
  - Node’s [`child_process`] says: “`'pipe'`: *Create a pipe* between
    the child process and the parent process”; this is the default,
    rather than “`'inherit'`: Pass through the corresponding stdio
    stream to/from the parent process” (emphasis mine).

[`Popen`]: https://docs.python.org/3/library/subprocess.html?highlight=subprocess#popen-constructor
[`child_process`]: https://nodejs.org/api/child_process.html#child_process_options_stdio

Reply:
Also—

> Yes, I don't care if TensorBoard can do it today or whatever. More complex system that spans multiple process is harder to introspect to understand that and that is the problem at stake.

—yep, that’s fair.

Reply:
Chatted offline. We think we’re talking about the same things: a Node
“stream” is a pipe with some associated machinery for handling events
and multicasting, and we’re using just the “pipe” part of that. Part of
the confusion was that `subprocess.PIPE` means “create new pipe (noun)”
rather than “pipe (verb) subprocess stdin to parent stdin”, which would
be “inherit” (`stdin=None` in Python).